ccm_web/server/src/api/api.create.section.handler.ts

```diff
@@ -0,0 +1,67 @@
+import CanvasRequestor from '@kth/canvas-api'
+
+import { CanvasSectionBase, CreateSectionsAPIErrorData, APIErrorData, isAPICreateSectionErrorData } from './api.interfaces'
+import { handleAPIError } from './api.utils'
+
+import baseLogger from '../logger'
+
+const logger = baseLogger.child({ filePath: __filename })
+
+export class CreateSectionApiHandler {
+  requestor: CanvasRequestor
+  sections: string[]
+  courseId: number
+
+  constructor (requestor: CanvasRequestor, sections: string[], courseId: number) {
+    this.requestor = requestor
+    this.sections = sections
+    this.courseId = courseId
+  }
+
+  async apiCreateSection (sectionName: string): Promise<CanvasSectionBase | CreateSectionsAPIErrorData> {
+    try {
+      const endpoint = `courses/${this.courseId}/sections`
+      const method = 'POST'
+      const requestBody = { course_section: { name: sectionName } }
+      logger.debug(`Sendings request to Canvas - Endpoint: ${endpoint}; Method: ${method}; Body: ${JSON.stringify(requestBody)}`)
+      const response = await this.requestor.requestUrl<CanvasSectionBase>(endpoint, method, requestBody)
+      const { id, name } = response.body
+      return { id, name }
+    } catch (error) {
+      const errResponse: APIErrorData = handleAPIError(error)
+      return { statusCode: errResponse.statusCode, errors: [{ message: errResponse.message, failedInput: sectionName }] }
+    }
+  }
+
+  makeReturnResponseCreateSections (sectionsReturnRes: Array<CreateSectionsAPIErrorData | CanvasSectionBase>): CanvasSectionBase[] | CreateSectionsAPIErrorData {
+    const successes = []; const statusCodes = []; const errorsList = []
+    for (const section of sectionsReturnRes) {
+      if (isAPICreateSectionErrorData(section)) {
+        const { statusCode, errors } = section
+        errorsList.push(...errors)
+        statusCodes.push(statusCode)
+      } else {
+        successes.push(section)
+      }
+    }
+
+    if (successes.length === this.sections.length) {
+      return successes
+    } else {
+      const uniqueStatusCodes = [...new Set(statusCodes)]
+      return {
+        statusCode: uniqueStatusCodes.length > 1 ? 400 : uniqueStatusCodes[0],
+        errors: errorsList
+      }
+    }
+  }
+
+  async createSectionBase (): Promise<CanvasSectionBase[] | CreateSectionsAPIErrorData> {
+    const start = process.hrtime()
+    const apiPromises = this.sections.map(async (section) => await this.apiCreateSection(section))
+    const sectionsOrErrorDataObjs = await Promise.all(apiPromises)
+    const stop = process.hrtime(start)
+    logger.debug(`Time taken to create ${this.sections.length} sections: ${(stop[0] * 1e9 + stop[1]) / 1e9} seconds`)
```

Comment:
It's probably unnecessary to compute the exact time spent on the process.  It's probably good enough to give begin/end debug log messages.  We can figure out from the timestamps in the log how long it took.

Reply:
I don't think it hurts to calculate, since I am not using any external logging library but using the Node built in library `process.hrtime`. I would like to keep it this way, But I am open for discussion if you have any concerns from that

Reply:
I agree.  Calculate the elapsed time all of the places!

Reply:
@chrisrrowland…
![Calculate ALL the elapsed times!](https://user-images.githubusercontent.com/20786/126799865-85811d3a-2974-4f88-b57b-ea88f093f145.jpg)